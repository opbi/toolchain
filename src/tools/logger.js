import pino from 'pino';

const devOrTestConfig = {
  prettyPrint: {
    colorize: true,
    translateTime: 'SYS:HH:MM:ss',
    ignore: 'pid,hostname',
    errorLikeObjectKeys: ['error'],
  },
};

const prodConfig = {
  // disable timestamp in prod to avoid conflict with fluentd
  timestamp: false,
  base: null,
  messageKey: 'message',
  useLevelLabels: true,
};

const devOrTestEnv = ['development', 'test'].includes(process.env.NODE_ENV);

export const config = {
  level: process.env.LOG_LEVEL || 'info',
  ...(devOrTestEnv ? devOrTestConfig : prodConfig),
};

const logger = pino(config);

export default logger;

// CONDITION: when the system needs to process sensitive data
// TODO: use pino-noir
// OUTCOME: mask sensitive data

// CONDITION: when adding flag system
// TODO: use pino-arborsculpture
// OUTCOME: control log level at runtime

// CONDITION: when creating debuggable lib modules
// TODO: config pino-debug
// OUTCOME: lib modules have their own debugger namespaces
