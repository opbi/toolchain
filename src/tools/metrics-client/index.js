import prometheus from 'prom-client';

import EnhancedCounter from './counter';
import Timer from './timer';

const { Summary, Gauge, Histogram } = prometheus;

export default {
  counter: EnhancedCounter,
  gauge: Gauge,
  histogram: Histogram,
  summary: Summary,
  timer: Timer,
};
