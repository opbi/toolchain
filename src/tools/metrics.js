import prometheus from 'prom-client';

import metricsClient from './metrics-client';

// TODO: update the original .inc(), .label() function
// to take only the labelNames[] from the input

// TODO: further abstract out type of metrics will be used on application
// e.g. error_counter, call_timer, etc.
// make metrics-specs to { action, type }?

// TODO: consider using fastify-metrics to get default histogram of response times

// TODO: need to test the behaviour in a distributed system

let metricsUnits = {};

/*
  when metrics were scraped by prometheus deamon,
  k8s meta data such as `app` would be added
  so it is not very necessary to add AppName into metrics namespace
  it can be used to distinguish with canned metrics
 */
const metricsNamespace = process.env.METRICS_NAMESPACE || 'custom_metrics';

/**
 * Add a new metrics unit from metrics-specs.
 *
 * @param {object} options - Config.
 * @param  {string}    options.type -  Type of the metrics, e.g. Counter.
 * @param  {string}    options.name -  Name of the metrics, e.g. Action_error.
 * @param  {object} options.specs - Other specific configuration for the metrics, e.g. Help, labelNames, buckets.
 */
const add = ({ type, name, ...specs }) => {
  const MetricsConstructor = metricsClient[type];

  if (!MetricsConstructor) {
    throw Error('invalid metrics type');
  }

  // metricsNamespace will be appended
  const m = new MetricsConstructor({
    ...specs,
    name: `${metricsNamespace}_${name}`,
  });

  metricsUnits[name] = m;
};

/**
 * Load a list of metrics specs to metricsUnits.
 *
 * @param  {Array} list - List of metrics specs.
 */
const load = (list) => {
  list.forEach((m) => {
    add(m);
  });
};

/**
 * List the metrics unit instances stored in metricsUnits.
 *
 * @returns {object} <{[metrics name]: metrics unit instance}>.
 */
const list = () => metricsUnits;

/**
 * Find a metrics unit by action/type or by name(higher priority).
 *
 * @param {object} options - Config.
 * @param  {string} options.action - Name of the action function.
 * @param  {string} options.type -   Type of metrics attached to the action function.
 * @param  {string} options.name -   Name of the metrics unit.
 * @returns {object}                The metrics unit instance.
 */
const find = ({ action, type, name }) => {
  const unitName = name || `${action}_${type}`;
  const unit = metricsUnits[unitName];
  if (!unit) {
    throw Error(`${unitName} not found in metrics-specs`);
  }
  return unit;
};

/**
 * Clear instances stored in metricsUnits and prometheus.register.
 */
const reset = () => {
  prometheus.register.clear();
  metricsUnits = {};
};

export default {
  add,
  load,
  list,
  find,
  reset,
};
