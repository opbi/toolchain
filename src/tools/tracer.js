import { initTracer } from 'jaeger-client';

import logger from './logger';

const config = {
  // serviceName is anchored to APP_NAME exported in Makefile
  serviceName: process.env.APP_NAME,
  sampler: {
    type: 'const',
    param: 1,
  },
  reporter: {
    logSpans: process.env.NODE_ENV === 'development',
  },
};
const option = { logger };

const configTracer = () => initTracer(config, option);

let instance;

const tracer = {
  rootSpan: undefined,
  start: (name) => {
    if (!instance) {
      instance = configTracer();
    }
    if (tracer.rootSpan) {
      const span = instance.startSpan(name, { childOf: tracer.rootSpan });
      return span;
    }
    const span = instance.startSpan(name);
    tracer.rootSpan = span;
    return span;
  },
};

export default tracer;
