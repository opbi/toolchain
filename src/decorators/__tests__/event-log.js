import { logger } from 'tools';

import eventLog from '../event-log';

logger.info = jest.fn();
logger.error = jest.fn();

describe('eventLog', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('log the success event with param when enabled', async () => {
    const original = async ({ text }) => text;
    const decorated = eventLog({ logParam: true })(original);
    const result = await decorated({ text: 'yes' }, {}, { logger });

    expect(result).toBe('yes');
    expect(logger.info.mock.calls).toMatchSnapshot();
  });

  it('log the success event with result when enabled', async () => {
    const original = async () => 'yes';
    const decorated = eventLog({ logResult: true })(original);
    const result = await decorated({}, {}, { logger });

    expect(result).toBe('yes');
    expect(logger.info.mock.calls).toMatchSnapshot();
  });

  it('log chained events from upper level function call', async () => {
    const original = () => {};
    const upper = (p, m, c) => {
      return eventLog()(original)(p, m, c);
    };
    await eventLog()(upper)({}, {}, { logger });
    expect(logger.info.mock.calls).toMatchSnapshot();
  });

  it('log the error event when input function fails', async () => {
    const original = () => {
      const e = { message: 'error' };
      throw e;
    };
    const decorated = eventLog()(original);

    try {
      await decorated({}, {}, { logger });
    } catch (e) {
      expect(e.message).toBe('error');
      expect(logger.info.mock.calls).toMatchSnapshot();
      expect(logger.error.mock.calls).toMatchSnapshot();
    }
  });

  it('log the parsed error when errorParse set', async () => {
    const original = () => {
      const e = new Error('error');
      throw e;
    };
    const decorated = eventLog({
      errorParser: e => ({
        message: e.message,
        stack: e.stack.substring(0, 20),
      }),
    })(original);

    try {
      await decorated({}, {}, { logger });
    } catch (e) {
      expect(e.message).toBe('error');
      expect(logger.info.mock.calls).toMatchSnapshot();
      expect(logger.error.mock.calls).toMatchSnapshot();
    }
  });
});
