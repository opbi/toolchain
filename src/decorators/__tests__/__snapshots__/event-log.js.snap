// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eventLog log chained events from upper level function call 1`] = `
Array [
  Array [
    Object {
      "event": "upper.original",
    },
    "upper.original",
  ],
  Array [
    Object {
      "event": "upper",
    },
    "upper",
  ],
]
`;

exports[`eventLog log the error event when input function fails 1`] = `Array []`;

exports[`eventLog log the error event when input function fails 2`] = `
Array [
  Array [
    Object {
      "error": Object {
        "message": "error",
      },
      "event": "original",
    },
    "error",
  ],
]
`;

exports[`eventLog log the parsed error when errorParse set 1`] = `Array []`;

exports[`eventLog log the parsed error when errorParse set 2`] = `
Array [
  Array [
    Object {
      "error": Object {
        "message": "error",
        "stack": "Error: error
    at ",
      },
      "event": "original",
    },
    "error",
  ],
]
`;

exports[`eventLog log the success event with param when enabled 1`] = `
Array [
  Array [
    Object {
      "event": "original",
      "param": Object {
        "text": "yes",
      },
    },
    "original",
  ],
]
`;

exports[`eventLog log the success event with result when enabled 1`] = `
Array [
  Array [
    Object {
      "event": "original",
      "result": "yes",
    },
    "original",
  ],
]
`;
